File: server/improvement/apply.js
// Applies and undoes self-improvement changes — rules and settings ONLY.
//
// THE INVARIANT THIS FILE EXISTS TO PROTECT: this module may never write to
// any file under the repo source tree, and never will — "Jarvis never
// edits its own code" is enforced here structurally, not just documented.
// Everything this file touches is a SQLite row (improvement-store.js) or a
// prefs.js key. A proposal whose kind is 'skill' or 'code' is refused
// outright by applyProposal() below, and improvement-policy.js's decide()
// already never routes either kind to auto-apply in the first place — this
// file is the second, independent gate, not the only one.

import * as store from './improvement-store.js';
import { getPrefs, setPrefs } from '../prefs.js';

/**
 * Prefs keys a self-improvement 'setting' proposal may ever touch — a
 * deliberate allowlist, not "anything in prefs.js." These are genuine
 * behaviour dials (how cautious, how fast, how much background work);
 * anything else in prefs.js (a pinned model id, the TTS provider, whether
 * auto-select itself is on) is structural configuration, not a behaviour
 * rule, and stays out of reach even at improvementTrust:'auto'.
 */
const ALLOWED_SETTING_KEYS = new Set(['balance', 'clarifySensitivity', 'memoryTrust', 'maxBackgroundJobs', 'improvementTrust', 'improvementResearch']);

// Every whitelisted key above is string-valued except this one — a caller
// this far from the actual pref shape (a model-called tool, in particular)
// may reasonably send "3" instead of 3. Coercing here, once, is cheaper and
// safer than trusting every caller to know prefs.js's exact type per key.
const NUMERIC_SETTING_KEYS = new Set(['maxBackgroundJobs']);

function coerceSettingValue(key, value) {
  if (NUMERIC_SETTING_KEYS.has(key)) {
    const n = Number(value);
    if (!Number.isFinite(n)) throw new Error(`"${value}" isn't a valid number for "${key}".`);
    return n;
  }
  return value;
}

function valuesEqual(a, b) {
  return JSON.stringify(a) === JSON.stringify(b);
}

/**
 * Applies a pending 'rule' or 'setting' proposal — called both by the
 * auto-apply path (once improvement-policy.js's decide() says
 * 'auto-apply') and by the screen's manual Approve button for anything a
 * human approves by hand, including a 'skill'/'code' idea the user
 * approves the CONCEPT of (which still never writes code here — see
 * implementation-prompt.js for what actually happens to those).
 *
 * Returns the created rule/updated pref value. Throws for any kind other
 * than 'rule'/'setting' — this function is never the path a Skill or code
 * change takes, regardless of who calls it.
 */
export function applyProposal(proposalId) {
  const proposal = store.getProposal(proposalId);
  if (!proposal) throw new Error('That suggestion no longer exists.');
  if (proposal.status !== 'pending') throw new Error(`That suggestion is already ${proposal.status}.`);

  if (proposal.kind === 'rule') {
    const text = String(proposal.payload?.text || '').trim();
    const scope = proposal.payload?.scope || 'general';
    if (!text) throw new Error('This proposal has no rule text to apply.');
    const rule = store.createRule({ text, scope, sourceProposalId: proposal.id });
    store.recordChange({
      kind: 'rule',
      target: rule.id,
      before: null,
      after: { text: rule.text, scope: rule.scope, active: rule.active },
      reason: proposal.rationale || 'Applied from a self-improvement suggestion.',
      proposalId: proposal.id,
    });
    store.setProposalStatus(proposal.id, 'applied');
    return rule;
  }

  if (proposal.kind === 'setting') {
    const key = proposal.payload?.key;
    if (!ALLOWED_SETTING_KEYS.has(key)) {
      throw new Error(`"${key}" isn't a setting self-improvement is allowed to change.`);
    }
    const value = coerceSettingValue(key, proposal.payload?.value);
    // Snapshot `before` and write `after` with NO await between them — the
    // one real hazard for a synchronous store like this: an interleaved
    // POST /api/prefs from the browser landing between a read and a write
    // would otherwise get silently clobbered. getPrefs()/setPrefs() are
    // both fully synchronous (store.js's readJson/writeJson), so this
    // block genuinely cannot yield to anything else mid-way.
    const before = getPrefs()[key];
    setPrefs({ [key]: value });
    store.recordChange({
      kind: 'setting',
      target: key,
      before,
      after: value,
      reason: proposal.rationale || 'Applied from a self-improvement suggestion.',
      proposalId: proposal.id,
    });
    store.setProposalStatus(proposal.id, 'applied');
    return { key, value };
  }

  throw new Error(`Self-improvement never applies a "${proposal.kind}" change directly — that always goes through the user first.`);
}

function currentValueFor(change) {
  if (change.kind === 'rule') {
    const rule = store.getRule(change.target);
    return rule ? { text: rule.text, scope: rule.scope, active: rule.active } : null;
  }
  if (change.kind === 'setting') {
    return getPrefs()[change.target];
  }
  return undefined; // 'undo' rows are never undoable — see undoChange() below
}

/** True only for a 'rule' change whose target no longer exists at all (hard-deleted from the archived view) — a genuinely different situation from "the user edited it since," which undoChange() must not conflate with the ordinary changed_since refusal (that one at least offers a real force option; this one has nothing left to restore). */
function targetWasDeleted(change) {
  return change.kind === 'rule' && !store.getRule(change.target);
}

/**
 * Undoes ONE applied change. Refuses instead of clobbering when the live
 * value no longer matches what this change last set (`after`) — the user
 * changed it themselves since, on the Settings screen or the
 * Self-Improvement screen's own rule text, and a blind restore would
 * silently overwrite that later, deliberate decision. Pass `force: true`
 * only after the caller has explicitly shown the user the mismatch and
 * they confirmed anyway.
 *
 * An 'undo' row (a change produced by a PREVIOUS call to this function) is
 * never itself undoable — re-applying something previously undone means
 * approving a fresh proposal, not "undoing an undo." This keeps the log
 * unambiguous: every row's `before`/`after` describes exactly what
 * happened at that moment, with no chain of reversals to reason through.
 *
 * Returns `{ok:true, restoredTo}` on success, `{ok:false, reason:
 * 'changed_since', expected, current}` when refusing because the live
 * value has moved on, or `{ok:false, reason:'target_deleted'}` when the
 * rule this change was about has since been hard-deleted entirely — a
 * genuinely different situation (nothing is left to restore, `force`
 * would silently no-op) that the screen must show differently.
 */
export function undoChange(changeId, { force = false } = {}) {
  const change = store.getChange(changeId);
  if (!change) throw new Error('That change no longer exists.');
  if (change.undoneAt) throw new Error('That change was already undone.');
  if (change.kind === 'undo') throw new Error("An undo can't itself be undone — approve a fresh suggestion to bring something back instead.");

  if (targetWasDeleted(change)) {
    return { ok: false, reason: 'target_deleted' };
  }

  const current = currentValueFor(change);
  if (!force && !valuesEqual(current, change.after)) {
    return { ok: false, reason: 'changed_since', expected: change.after, current };
  }

  let restoredTo;
  if (change.kind === 'rule') {
    // The only kind of rule change apply.js writes with before:null is a
    // CREATION — undoing one means deactivating it, never a hard delete,
    // so its own history stays intact and answerable (same choice
    // memory-store.js's mergeMemories() makes: archive, never delete).
    // A non-null `before` means an EDIT (server.js's rule-edit route
    // records one via improvement-store.js's updateRuleText()) — undoing
    // THAT means putting the wording back, not just flipping active.
    if (change.before == null) {
      store.setRuleActive(change.target, false);
      restoredTo = { ...currentValueFor(change) };
    } else {
      const before = change.before;
      store.updateRuleText(change.target, before.text);
      if (typeof before.active === 'boolean') store.setRuleActive(change.target, before.active);
      restoredTo = { ...currentValueFor(change) };
    }
  } else if (change.kind === 'setting') {
    setPrefs({ [change.target]: change.before });
    restoredTo = change.before;
  } else {
    throw new Error(`Don't know how to undo a "${change.kind}" change.`);
  }

  store.markChangeUndone(changeId);
  store.recordChange({
    kind: 'undo',
    target: change.target,
    before: change.after,
    after: restoredTo,
    reason: `Undo of an earlier ${change.kind} change.`,
    proposalId: change.proposalId,
  });

  return { ok: true, restoredTo };
}

File: backend/jarvis/improvement/apply.py
"""The only thing that changes anything, and the only thing that undoes it.

**Jarvis never edits its own code — structurally, not by convention.** This
module writes exactly two things: rows in the improvement tables, and keys in
preferences. It cannot write a file under the source tree because it never opens
one, and it refuses outright for any proposal kind other than a rule or a
setting. A proposal that needs real code produces a BRIEF for a human's coding
assistant instead; generating that brief IS the approval action for those kinds.

**Undo refuses rather than clobbers.** It compares the live value against what
the change actually set. If the user has since changed it themselves — muted the
rule, edited the preference — restoring the old value would silently overwrite
their own later decision, so it says so and stops unless told to proceed anyway.
An undo is itself never undoable: bringing something back means approving it
again, which is what keeps every before/after pair in the log honest.
"""

from __future__ import annotations

from typing import Any

from ..prefs import get_prefs, set_prefs
from . import store
from .policy import AUTO_APPLY, APPLIABLE_KINDS, decide


class UndoRefused(RuntimeError):
    """The live value is not what this change set. Carries what it found."""

    def __init__(self, message: str, *, live: Any, expected: Any):
        super().__init__(message)
        self.live = live
        self.expected = expected


def apply_proposal(proposal_id: str, *, approved_by: str = "user") -> dict[str, Any]:
    """Make the change a proposal describes. Refuses anything that is not a rule
    or a setting, whatever the caller believes it has approved."""
    proposal = store.get_proposal(proposal_id)
    if proposal is None:
        raise KeyError(f"No such proposal: {proposal_id}")
    if proposal["kind"] not in APPLIABLE_KINDS:
        raise ValueError(
            f"A {proposal['kind']} proposal is never applied automatically — it needs "
            "real work by a person.")

    payload = proposal.get("payload") or {}
    if proposal["kind"] == "rule":
        rule = store.create_rule(text=str(payload.get("text") or proposal["title"]),
                                 scope=str(payload.get("scope") or "general"),
                                 source_proposal_id=proposal_id)
        change = store.record_change(kind="rule", target=rule["id"], before=None,
                                     after={"text": rule["text"], "active": True},
                                     reason=f"approved by {approved_by}",
                                     proposal_id=proposal_id)
    else:
        key = str(payload.get("key") or "")
        if not key:
            raise ValueError("A setting proposal has to say which setting.")
        before = get_prefs().get(key)
        after = payload.get("value")
        set_prefs({key: after})
        change = store.record_change(kind="setting", target=key, before=before, after=after,
                                     reason=f"approved by {approved_by}",
                                     proposal_id=proposal_id)

    store.set_proposal_status(proposal_id, "approved")
    return change


def auto_apply_if_allowed(proposal_id: str) -> dict[str, Any] | None:
    """Apply only if the policy says so. The policy is asked here rather than
    trusted from the caller, so a caller that forgot to check cannot bypass it."""
    proposal = store.get_proposal(proposal_id)
    if proposal is None:
        return None
    if decide(proposal) != AUTO_APPLY:
        return None
    return apply_proposal(proposal_id, approved_by="Jarvis (within your settings)")


def undo_change(change_id: int | str, *, force: bool = False) -> dict[str, Any]:
    change = store.get_change(change_id)
    if change is None:
        raise KeyError(f"No such change: {change_id}")
    if change.get("undone_at"):
        raise ValueError("That has already been undone.")
    if change["kind"] == "undo":
        # Bringing something back means approving it again, not reversing a
        # reversal — which is what keeps the log's before/after pairs honest.
        raise ValueError("An undo can't itself be undone.")

    if change["kind"] == "rule":
        rule = store.get_rule(change["target"])
        if rule is None:
            raise ValueError("That rule is no longer here.")
        live = {"text": rule["text"], "active": bool(rule["active"])}
        if not force and live != (change.get("after") or {}):
            raise UndoRefused(
                "That rule has changed since — undoing would overwrite what you did to it.",
                live=live, expected=change.get("after"))
        store.archive_rule(rule["id"])
        store.set_rule_active(rule["id"], False)
    else:
        key = change["target"]
        live = get_prefs().get(key)
        if not force and live != change.get("after"):
            raise UndoRefused(
                f"'{key}' has changed since — undoing would overwrite what you set.",
                live=live, expected=change.get("after"))
        set_prefs({key: change.get("before")})

    store.mark_change_undone(change_id)
    return store.record_change(kind="undo", target=change["target"],
                               before=change.get("after"), after=change.get("before"),
                               reason=f"undo of {change_id}")
